ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/LedgerDB/V1/BackingStore/Mock.hs

```diff
@@ -0,0 +1,338 @@
+
+import           Control.Monad
+import           Control.Monad.Except (ExceptT (..), MonadError (throwError),
+                     catchError, runExceptT)
+import           Control.Monad.State (MonadState, State, StateT (StateT), gets,
+                     modify, runState)
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Ouroboros.Consensus.Block.Abstract (SlotNo, WithOrigin (..))
+import qualified Ouroboros.Consensus.Storage.LedgerDB.V1.BackingStore as BS
+import qualified System.FS.API.Types as FS
+
+{-------------------------------------------------------------------------------
+  Types
+-------------------------------------------------------------------------------}
+
+data Mock vs = Mock {
+    backingValues :: vs
+  , backingSeqNo  :: WithOrigin SlotNo
+  , copies        :: Map FS.FsPath (WithOrigin SlotNo, vs)
+  , isClosed      :: Bool
+    -- | Track whether value handles have been closed.
+  , valueHandles  :: Map ID ValueHandleStatus
+    -- | The next id to use if a new value handle is opened.
+  , nextId        :: ID
+  }
+  deriving stock (Show, Eq)
+
+data ValueHandleStatus = Open | ClosedByStore | ClosedByHandle
+  deriving stock (Show, Eq)
+
+data ValueHandle values = ValueHandle {
+    getId  :: ID
+  , values :: values
+  , seqNo  :: WithOrigin SlotNo
+  }
+  deriving stock Show
+
+instance Eq (ValueHandle vs) where
+  x == y = getId x == getId y
+
+instance Ord (ValueHandle vs) where
+  x <= y = getId x < getId y
+
+-- | An ID for a mocked value handle.
+newtype ID = ID Word
+  deriving stock (Show, Eq, Ord)
+  deriving newtype Num
+
+-- | An empty mock state.
+emptyMock :: EmptyValues vs => Mock vs
+emptyMock = Mock {
+    backingValues = emptyValues
+  , backingSeqNo  = Origin
+  , copies        = Map.empty
+  , isClosed      = False
+  , valueHandles  = Map.empty
+  , nextId        = 0
+  }
+
+data Err =
+    ErrBackingStoreClosed
+  | ErrBackingStoreValueHandleClosed
+  | ErrCopyPathAlreadyExists
+  | ErrCopyPathDoesNotExist
+  | ErrNonMonotonicSeqNo (WithOrigin SlotNo) (WithOrigin SlotNo)
+  deriving stock (Show, Eq)
+
+{-------------------------------------------------------------------------------
+  Type classes
+-------------------------------------------------------------------------------}
+
+-- | Abstract over interactions between values, keys and diffs.
+class ( EmptyValues vs, ApplyDiff vs d, LookupKeysRange ks vs
+      , LookupKeys ks vs, ValuesLength vs, MakeDiff vs d
+      , DiffSize d, KeysSize ks
+      ) => HasOps ks vs d
+
+class EmptyValues vs where
+  emptyValues :: vs
+
+class ApplyDiff vs d where
+  applyDiff :: vs -> d -> vs
+
+class LookupKeysRange ks vs where
+  lookupKeysRange :: Maybe ks -> Int -> vs -> vs
+
+class LookupKeys ks vs where
+  lookupKeys :: ks -> vs -> vs
+
+class ValuesLength vs where
+  valuesLength :: vs -> Int
+
+class MakeDiff vs d where
+  diff :: vs -> vs -> d
+
+class DiffSize d where
+  diffSize :: d -> Int
```

Comment:
What is this counting? Does it count 0, 1, or 2 if a diff adds a key and also deletes that key, etc?